WordPress/gutenberg
# Gutenberg: Add a HMR & React Fast Refresh development mode.

## Description

Closes #23013.

### Background

Webpack Hot Module Replacement (HMR) is fundamentally at odds with the way we do script overriding in Gutenberg.

After lots of experimentation and failed attempts, I found that the best way to support it is to provide an alternative development mode on a per-page/app basis. For example, the site editor is its own page/app, and this PR enables an HMR mode for its development.

It should be trivial to reuse the same configurations for the post, widgets, and navigation editors.

I also included the new [React Fast Refresh](https://reactnative.dev/docs/fast-refresh) experience. The developer experience with this is fantastic. Even error resilience and recovery are flawless, both for runtime and syntax errors. It's nice to be able to iterate on a screen without resetting state.

### Implementation

This PR introduces a new Webpack config for running the new dev server as an extension of the one provided by `@wordpress/scripts`.

Pages that wish to support an HMR mode, need to create a `hot.js` entry point like this PR does for `edit-site`. This entry point should initialize the page and load its root stylesheet.

On the server, we use a new config global, `GUTENBERG_HMR`, to determine whether to load the new script from the dev server or not. We manipulate this config variable effortlessly in the scripts introduced by this PR by using `wp-env`. It will enable it before running the new development mode, and disable it before running the old one. This makes it easy to switch between the two, but it does assume you are using `wp-env` to manage your environment.

## How to test this?

- Set up your environment using [`wp-env`](https://www.npmjs.com/package/@wordpress/env) if you haven't yet.
- Run `npm run dev:edit-site`.
- Go to the site editor and see how it updates in response to your code changes without reloading.

## Screenshots

![gif](https://user-images.githubusercontent.com/19157096/84104016-7e420d00-a9c9-11ea-8d64-0b55d2fb1493.gif)

## Checklist:
